File: kotlin/src/main/kotlin/com/cyaim/im/client/Backoff.kt
package com.cyaim.im.client

import kotlin.random.Random
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds

/**
 * Reconnect backoff: exponential ceiling, **full jitter**.
 *
 * A gateway holds sockets, so it deploys rarely — but when it does, every client on the node being
 * replaced is disconnected in the same instant, and every one of them wants back in. The delay a
 * client waits is the only thing standing between that and a synchronised stampede against the
 * service that has just finished starting.
 *
 * Full jitter means a uniform draw from `[0, ceiling)`, where the ceiling doubles from 500 ms up
 * to 30 s. Not a fixed backoff, and not `delay ± 10 %`: with either of those, N clients that were
 * dropped together arrive together, one ceiling later, and knock the node over again. Only drawing
 * across the *whole* window actually spreads them out — after the first round the arrivals are
 * uniform over 30 seconds instead of stacked in one.
 *
 * 满抖动（0..ceiling 均匀取值），不是固定退避、也不是 ±10%：网关滚动更新时同一节点上的连接
 * 会被同时断开，只有在整个窗口上均匀取值才能把它们真正打散。
 */
public object Backoff {
    /** First ceiling. Short enough that a one-off blip is invisible to the user. */
    public const val BASE_DELAY_MS: Long = 500

    /**
     * Hard ceiling. Also the reason a phone that comes back from Doze reconnects within half a
     * minute instead of an hour: the window never grows past this, however long we were away.
     */
    public const val MAX_DELAY_MS: Long = 30_000

    /** 500 ms shifted six times is 32 s, already past [MAX_DELAY_MS]; shifting further is noise. */
    public const val MAX_DOUBLINGS: Int = 6

    /** The upper bound of the window for [attempt] (1 = the first retry). */
    public fun ceilingMillis(attempt: Int): Long =
        minOf(MAX_DELAY_MS, BASE_DELAY_MS shl attempt.coerceIn(0, MAX_DOUBLINGS))

    /**
     * A delay drawn uniformly from `[0, ceilingMillis(attempt))`.
     *
     * [random] is a parameter so the distribution can be asserted in a test. A reconnect policy
     * that is only inspected by reading it is a reconnect policy that quietly regresses to a fixed
     * delay the first time someone "simplifies" it.
     */
    public fun nextDelay(attempt: Int, random: Random = Random.Default): Duration =
        random.nextLong(ceilingMillis(attempt)).milliseconds
}

File: swift/Sources/CyaimIM/Backoff.swift
import Foundation

/// Reconnect delays, drawn with **full jitter**.
///
/// The ceiling doubles per attempt up to 30 seconds, and the delay actually used is a uniform draw
/// from `0 ..< ceiling`. Not the ceiling itself, and not the ceiling ±10%.
///
/// This matters more than it looks. A gateway holds sockets, so a rolling update drops every socket
/// on the node being replaced at the same instant — tens of thousands of clients, all of which now
/// want to reconnect. With a fixed backoff they come back in the same second, at the service that
/// has only just finished starting, and knock it over again; the second wave is worse than the
/// first because it is now synchronised. Jittering by ±10% barely helps: the arrivals are still a
/// spike, just a slightly wider one. Drawing uniformly from the whole interval is what actually
/// spreads a node's worth of clients across it.
///
/// 满抖动而不是固定退避：网关滚动更新会让整节点客户端同时断线，固定退避（或 ±10% 抖动）
/// 会让它们在同一秒回来，把刚恢复的服务再打垮一次。
public struct FullJitterBackoff: Sendable, Hashable {
    /// Ceiling for the first attempt. Doubles from here.
    public var initialDelay: Duration

    /// The ceiling stops growing here. Beyond ~30s a user has given up on the app anyway, and the
    /// spread is already wide enough to de-synchronise a fleet.
    public var maximumDelay: Duration

    /// How many doublings before the ceiling is clamped. Six doublings of 500ms overshoots 30s,
    /// which is the intent: attempts 6 and beyond all draw from the full 0..30s window.
    public var doublingLimit: Int

    public init(
        initialDelay: Duration = .milliseconds(500),
        maximumDelay: Duration = .seconds(30),
        doublingLimit: Int = 6
    ) {
        self.initialDelay = initialDelay
        self.maximumDelay = maximumDelay
        self.doublingLimit = doublingLimit
    }

    public static let `default` = FullJitterBackoff()

    /// The upper bound for `attempt`, where the first reconnect after a drop is attempt 1.
    public func ceiling(forAttempt attempt: Int) -> Duration {
        let exponent = min(max(attempt, 1), doublingLimit)
        let scaled = initialDelay.seconds * pow(2, Double(exponent))
        return .seconds(min(maximumDelay.seconds, scaled))
    }

    /// A delay for `attempt`, drawn uniformly from `0 ..< ceiling(forAttempt:)`.
    public func delay(forAttempt attempt: Int) -> Duration {
        var generator = SystemRandomNumberGenerator()
        return delay(forAttempt: attempt, using: &generator)
    }

    /// A delay for `attempt`, drawn from an explicit generator so the distribution is testable.
    public func delay<Generator: RandomNumberGenerator>(
        forAttempt attempt: Int,
        using generator: inout Generator
    ) -> Duration {
        let ceilingSeconds = ceiling(forAttempt: attempt).seconds
        guard ceilingSeconds > 0 else { return .zero }
        return .seconds(Double.random(in: 0 ..< ceilingSeconds, using: &generator))
    }
}

extension Duration {
    /// This duration in seconds. `Duration` stores attoseconds, and the arithmetic below is all in
    /// fractional seconds, so the conversion lives in one place.
    var seconds: Double {
        let parts = components
        return Double(parts.seconds) + Double(parts.attoseconds) / 1_000_000_000_000_000_000
    }
}
